### ISSUE TITLE
[Regression] background_jobs segment doesn't update in next branch when jobs finish

### ISSUE THREAD

**Issue by romkatv** [bhilburn/powerlevel9k#1196]
In `release` branch, when a background job completes, `background_jobs` segment updates. In `next` it doesn't.

Here's `master`:

```zsh
docker run -e LANG=C.UTF-8 -e LC_ALL=C.UTF-8 -e TERM=$TERM -it --rm ubuntu bash -uexc '
  apt update && apt install -y zsh git
  git clone -b master https://github.com/bhilburn/powerlevel9k.git ~/powerlevel9k
  echo "
    POWERLEVEL9K_LEFT_PROMPT_ELEMENTS=()
    POWERLEVEL9K_RIGHT_PROMPT_ELEMENTS=(background_jobs)
    source ~/powerlevel9k/powerlevel9k.zsh-theme
    sleep 3&" >~/.zshrc
  zsh -i'
```

![job-complete-master](https://user-images.githubusercontent.com/1282067/54357558-a3c55700-465e-11e9-83e0-1d27c472d7f8.png)

Replace `master` with `next` to test with `next` branch.

![job-complete-next](https://user-images.githubusercontent.com/1282067/54357583-acb62880-465e-11e9-8823-557475576512.png)

Notice how `next` is still saying there is a background job after the job completes.

**Comment by Syphdias** [comment]
Hm, I'm not quite sure how to circumvent this. This happens because `__p9k_prepare_prompts` is not triggered after the `notify` "done" before the prompt is redrawn. Is dosn't update `__p9k_unsafe_prompt` which is evaluated from `PROMPT`.

1. Using `%j` would yield the correct job number at least but compromises on being extendable for splitting up jobs (running + suspended). Also, the segment would still be there and not hidden for one prompt. This could be used if the "expand" option is not set, but is only a half-way fix.
2. Maybe there is some kind of hook for the "done" notify to run the precmd hook/`__p9k_prepare_prompts` function before rendering.
3. Another fix would be to revert the changes, which would lead to the same problems as before in #1119 
4. For completeness sake of my brainstorm: You could do `unsetopt notify`. This would prevent the prompt rerender since "done" is never shown, but would compromise user settings, which I do not like at all!

Side note: After hitting enter you get the correct prompt

@dritter @romkatv any better ideas?

**Comment by romkatv** [comment]
It works in Powrelevel10k. You can take a look how it's done. It's really brutal though. And because of a bug in zsh the prompt refreshes only after a second. Still better than no refresh at all. I'll post a link to the bug once I send the report to zsh-workers. If you find a better solution, let me know.

I've also started thinking about a very different approach to rendering segments that would solve this and other problems. The ideas are still very raw. I need to educate myself better about zsh before I can flesh them out.

**Comment by Syphdias** [comment]
It works in powerlevel10k because your `PROMPT` is probably more similar to the one on `master` where the build function is called every time the `PROMPT` variable is evaluated. The bug described in #1119 only triggers if the prompt render is interrupted (e.g. by holding down `ctrl-c`). Then the unevaluated `PROMPT` is shown. At least that is what I think is happening.

For Powerlevel10k you probably don't run into this because `ctrl-c` is probably not fast enough to interrupt rendering. I didn't test this hypothesis though.

**Comment by romkatv** [comment]
> It works in powerlevel10k because your PROMPT is probably more similar to the one on master where the build function is called every time the PROMPT variable is evaluated.

Ha-ha, no :-D

Powerlevel10k only looks the same. The code is almost 100% rewritten (except for the odd segments that I don't use).

**Comment by romkatv** [comment]
How do you think I knew that `background_jobs` is probably broken in `next`? You told me you've removed process substitution from prompt. I've already done it and knew that it breaks `background_jobs`, so I assumed it broke in `next` too.

**Comment by romkatv** [comment]
I opened another bug: https://github.com/bhilburn/powerlevel9k/issues/1201. Figured you'll probably have to go through everything that stops working once you remove process substitution, so I might as well save you time.

**Comment by romkatv** [comment]
I just finished a new implementation of background_jobs in Powerlevel10k that passes all the checks. No process substitutions, super fast, and updates instantly when jobs finish. Feel free to peek. It's implemented on top of the new rendering engine for segments I built, which is incredibly powerful but also incredibly complex.

**Comment by Syphdias** [comment]
I didn't find a hook that gets triggered after a background job is done. Which means I can't run the necessary `__p9k_prepare_prompts` function before prompt render like it is possible for [`vi_mode`](https://github.com/bhilburn/powerlevel9k/issues/1201#issuecomment-475890735).

So I see two ways to tackle this.
1. Add a "timer" for every instance of zsh to check if the number of background jobs changed and then trigger `__p9k_prepare_prompts` and `reset-prompt` if different.
Pros:
\- Could also be used for different things like #1190 / #1193 
Cons:
\- Extra process for every zsh instance (probably wouldn't need much computing power)
\- The prompt would render twice after a background_job done.
\- The second render would be delayed in average by half the delay the timer has (so .5s probably)
2. Looking at @romkatv's implementation. He does two things differently, as far as I understood his code:
a. instead of working with `echo -n` he uses a global variable to build his prompt: `_P9K_PROMPT+="new content from segment"`. This takes care that there is not `ctrl-c` bug. (Not sure how this handles not expanding evil code in prompt but apparently it doesn't run it.)
b. I think he adds something to not render the segment if it returns empty even if the precmd function was not called. https://github.com/romkatv/powerlevel10k/blob/0710cc7221cef20b73eeb75c6b3f2043ff56d019/powerlevel9k.zsh-theme#L226? I makes my brain melt... While this expansion stuff is what seems to tickle @romkatv in the right ways, it does not for me. :sweat_smile: I brooded for 2-3 hours and gave up now...

@dritter, thoughts?

**Comment by dritter** [comment]
IMHO we should use `%j` as @romkatv does. And in the extended case a subshell. This is not ideal, but seems to be the best compromise in terms of code readability and speed.
A timer and/or a precmd hook look too complicated to me.

**Comment by romkatv** [comment]
`echo` vs `_P9K_PROMPT+="new content from segment"` is a red herring. You get problems with Ctrl-C when you have command substitution in `PROMPT`. It's a bad idea to put command substitution in `PROMPT` not just for this reason but even this one is enough.

Powerlevel10k doesn't put command substitution in `PROMPT`, hence you cannot glitch it with Ctrl-C. How `PROMPT` is actually built is irrelevant. You can use `echo` and command substitutions in `precmd` all your want -- that stuff won't spill its guys like `PROMPT` does when users smash Ctrl-C.

> Not sure how this handles not expanding evil code in prompt but apparently it doesn't run it.

You mean not doing expansion on things like directory names? Powerlevel10k does this better than Powerlevel9k. See https://github.com/bhilburn/powerlevel9k/issues/1206 and several other bugs I opened.

> IMHO we should use %j as @romkatv does. And in the extended case a subshell.

Using `%j` is a good idea because it's much faster. Subshell is tricky though because you need to handle the case when the last background job disappears and you need to remove the whole segment. You cannot do it by using command substitution just for the background jobs segment because connection colors in the neighboring segments will need to change. So you'll have to wrap the whole prompt in `$(build_prompt)` again, which gets you the nasty Ctrl-C issue back, no realtime clock, and bad performance.

**Comment by romkatv** [comment]
> I think he adds something to not render the segment if it returns empty even if the precmd function was not called.

That's right. Segments in Powerlevel10k have dynamic conditions attached to them. When `PROMPT` is expanded, segments whose conditions expand to null are skipped over. Expansion of `PROMPT` has a state machine in it, which allows the remaining segments to get rendered with the right connections.

Check out https://www.reddit.com/r/zsh/comments/b1xxuj/abusing_zsh_parameter_expansion_for_fun_and/eiqeauc?utm_source=share&utm_medium=web2x, where I explain how this machinery works. There is a detailed example, too.

This code will get simpler once I finish the extension API.

**Comment by Syphdias** [comment]
TL;DR of the last 3 posts: I can't do it like this without re-implementing code render as you did; `%j` should be good enough for now

**Comment by dritter** [comment]
Closing this issue for now, as the discussion proceeded in #1254 and that one got merged.
